File: python/53.最大子数组和.py
#
# @lc app=leetcode.cn id=53 lang=python3
#
# [53] 最大子数组和
#

# @lc code=start
class Solution:

    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)
        ans = [0] * n
        ans[0] = nums[0]
        ret = ans[0]
        for i in range(1, n):
            ans[i] = max(ans[i - 1] + nums[i], nums[i])
            ret = max(ret, ans[i])
        return ret
# @lc code=end

File: cpp/53.最大子数组和.cpp
/*
 * @lc app=leetcode.cn id=53 lang=cpp
 *
 * [53] 最大子数组和
 */
#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
// @lc code=start
class Solution
{
public:
    int maxSubArray(vector<int>& nums)
    {
        return method1(nums);
    }

    // dp解法
    int method1(vector<int>& nums)
    {
        int pre = 0;
        int res = nums[0];
        for (auto num : nums) {
            pre = max(pre + num, num);
            res = max(pre, res);
        }
        return res;
    }

    // TODO 分治解法
};
// @lc code=end
